kani-compiler/src/codegen_cprover_gotoc/codegen/assert.rs

```diff
@@ -2,35 +2,80 @@
 
 //! This module is the central location for handling assertions and assumptions in Kani.
+//!
+//! There are a few patterns we see with CBMC:
+//!
+//! 1. A Kani `check` is a CBMC `assert`, which allows execution to proceed if it fails.
+//! 2. A Kani `assume` is a CBMC `assume`, thankfully.
+//! 3. A Kani `assert` is a CBMC `assert-assume`, first checking the property, then terminating the trace if it fails.
+//!
+//! Kani further offers a few special cases:
+//!
+//! 4. `codegen_unimplemented` : `assert(false)` but recorded specially
+//! 5. `codegen_sanity` : `assert` but not normally displayed as failure would be a Kani bug
+//!
 
 use crate::codegen_cprover_gotoc::utils;
 use crate::codegen_cprover_gotoc::GotocCtx;
-use cbmc::goto_program::{BuiltinFn, Expr, Location, Stmt};
+use cbmc::goto_program::{Expr, Location, Stmt, Type};
+use cbmc::InternedString;
 use rustc_span::Span;
 use std::convert::AsRef;
 use strum_macros::{AsRefStr, EnumString};
+use tracing::debug;
 
-/// The Property Class enum stores all viable options for classifying asserts, cover assume and other related statements
+/// Classifies the type of CBMC `assert`, as different assertions can have different semantics (e.g. cover)
+///
+/// Each property class should justify its existence with a note about the special handling it recieves.
 #[derive(Debug, Clone, EnumString, AsRefStr)]
 #[strum(serialize_all = "snake_case")]
 pub enum PropertyClass {
+    /// Overflow panics that can be generated by Intrisics.
+    /// NOTE: Not all uses of this are found by rust-analyzer because of the use of macros. Try grep instead.
+    ///
+    /// SPECIAL BEHAVIOR: None TODO: Why should this exist?
     ArithmeticOverflow,
+    /// The Rust `assume` instrinsic is `assert`'d by Kani, and gets this property class.
+    ///
+    /// SPECIAL BEHAVIOR: None? Possibly confusing to customers that a Rust assume is a Kani assert.
     Assume,
+    /// See [GotocCtx::codegen_cover] below. Generally just an `assert(false)` that's not an error.
+    ///
+    /// SPECIAL BEHAVIOR: "Errors" for this type of assertion just mean "reachable" not failure.
     Cover,
-    /// Assertions and Panic that are not specific to Kani. In a concrete execution, we expect
-    /// these assertions to be available to the user.
-    /// E.g.: User assertions, compiler invariant checks
+    /// Ordinary (Rust) assertions and panics.
+    ///
+    /// SPECIAL BEHAVIOR: None, this is the default type of assertion.
     Assertion,
+    /// Another instrinsic check.
+    ///
+    /// SPECIAL BEHAVIORL None TODO: Why should this exist?
     ExactDiv,
+    /// The `kani::expect_fail` assertion.
+    ///
+    /// SPECIAL BEHAVIOR: Inverted interpretation of success/failure.
+    /// (Note: Possibly this should be the same as Cover, or perhaps an ExplicitCover?)
     ExpectFail,
+    /// Another instrinsic check.
+    ///
+    /// SPECIAL BEHAVIOR: None TODO: Why should this exist?
```

Comment:
This is another instance of a check that we include to detect undefined behavior coming from intrinsics. I think that's the purpose of `SafetyCheck` (i.e., checks added by Kani to detect the violation of safety conditions), so we could use that instead.